File: leetcode/leetcode_cpp/count-good-nodes-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 /*
 https://leetcode.com/problems/count-good-nodes-in-binary-tree
 
 1. DFS preorder
 time: o(n)
 space: o(n)
 */
class Solution {
public:
    int res = 0;
    void dfs(const TreeNode* root, int cur) {
        if (!root) return;
        if (root->val >= cur) {
            res += 1;
            cur = root->val;
        }
        dfs(root->left, cur);
        dfs(root->right, cur);
    }
    int goodNodes(TreeNode* root) {
        if (!root) return 0;
        dfs(root, root->val);
        return res;
    }
};

File: leetcode/leetcode_py/count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
'''
https://leetcode.com/problems/count-good-nodes-in-binary-tree
 
1. DFS preorder
time: o(n)
space: o(n)
'''
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if not root:
            return 0
        def dfs(node, val):
            if not node:
                return 0
            cnt = 0
            if node.val >= val:
                cnt = 1
                val = node.val
            cnt += dfs(node.left, val)
            cnt += dfs(node.right, val)
            return cnt
        return dfs(root, root.val)